File: modeling/orbits.py
import csv
import os

# Orbit configs live in orbits.csv (shared with orbits.c), not here -- add a
# new orbit by appending rows there (same name on each of its particle rows),
# not by editing this file.
_CSV_PATH = os.path.join(os.path.dirname(__file__), "orbits.csv")


def _load_configs():
    configs = {}
    with open(_CSV_PATH, newline="") as f:
        for row in csv.DictReader(f):
            cfg = configs.setdefault(row["name"], {"n": 0, "dt": float(row["dt"]), "particles": []})
            cfg["particles"].append({k: float(row[k]) for k in ("rx", "ry", "rz", "vx", "vy", "vz", "m")})
            cfg["n"] = len(cfg["particles"])
    return configs


_configs = _load_configs()
# CONFIGS[name] is a zero-arg callable (not the dict directly) to match the
# pre-CSV interface, where each entry computed its particle data on call.
CONFIGS = {name: (lambda cfg=cfg: cfg) for name, cfg in _configs.items()}

File: modeling/orbits.c
#include <string.h>
#include <stdlib.h>
#include <stdio.h>

#include "orbits.h"

// Orbit configs live in orbits.csv (shared with orbits.py), not here -- add
// a new orbit by appending rows there (same name on each of its particle
// rows), not by editing this file.
#define ORBITS_CSV_PATH "modeling/orbits.csv"

static inline int32_t to_fixed(double x) {
    return (int32_t)llround(x * (double)FRAC_SCALE);
}

double load_orbit_config(const char *name, State *state) {
    FILE *f = fopen(ORBITS_CSV_PATH, "r");
    if (!f) {
        perror("fopen " ORBITS_CSV_PATH);
        exit(1);
    }

    char line[256];
    fgets(line, sizeof(line), f);  // skip header row

    memset(state, 0, sizeof(*state));
    int n = 0;
    double dt = 0.0;
    int found = 0;

    char row_name[64];
    double row_dt, rx, ry, rz, vx, vy, vz, m;
    while (fgets(line, sizeof(line), f)) {
        if (sscanf(line, "%63[^,],%lf,%lf,%lf,%lf,%lf,%lf,%lf,%lf",
                   row_name, &row_dt, &rx, &ry, &rz, &vx, &vy, &vz, &m) != 9) {
            continue;
        }
        if (strcmp(row_name, name) != 0) continue;

        found = 1;
        dt = row_dt;
        if (n >= N) {
            fprintf(stderr, "orbit '%s' needs more than %d bodies\n", name, N);
            exit(1);
        }
        state->rx[n] = to_fixed(rx);
        state->ry[n] = to_fixed(ry);
        state->rz[n] = to_fixed(rz);
        state->vx[n] = to_fixed(vx);
        state->vy[n] = to_fixed(vy);
        state->vz[n] = to_fixed(vz);
        state->m[n]  = to_fixed(m);
        n++;
    }
    fclose(f);

    if (!found) {
        fprintf(stderr, "unknown orbit '%s'\n", name);
        exit(1);
    }

    return dt;
}
